### PR TITLE
Use http-client-0.5.0, http-conduit-2.2.0, http-client-tls-0.3.0

### PR DESCRIPTION
**Pull request by sjakobi**
Because I'm not really familiar with `http-client` I usually opted to replace the deprecated `parseUrl` with the synonymous `parseUrlThrow`.

Closes https://github.com/commercialhaskell/stack/issues/2332.

### CODE REVIEW

**Comment by borsboom** [comment]
I'd just gotten started with this process when I saw your PR come through, and looks like you took the same approach I'd begun, so great!

The only thing I was thinking of doing differently was that in cases where a `checkStatus = \_ _ _ -> Nothing` was set, I instead replaced `parseUrl` with `parseRequest` and removed the `checkStatus` (rather than changing to `parseUrlThrow` and `checkResponse`).  I _think_ that was safe, but I'm not 100% certain.

File: src/Stack/Upload.hs [hunk 1]
```diff
@@ -195,10 +195,9 @@ mkUploader config us = do
     manager <- usGetManager us
     (creds, fromFile') <- loadCreds $ usCredsSource us config
     when (not fromFile' && usSaveCreds us) $ saveCreds config creds
-    req0 <- parseUrl $ usUploadUrl us
+    req0 <- parseRequest $ usUploadUrl us
     let req1 = req0
             { requestHeaders = [("Accept", "text/plain")]
-            , checkStatus = \_ _ _ -> Nothing
```
Comment (sjakobi) [inline src/Stack/Upload.hs]: @borsboom:  I did the same in this case.

File: src/Network/HTTP/Download.hs [hunk 2]
```diff
@@ -108,7 +108,7 @@ redownload req0 dest = do
                         requestHeaders req0 ++
                         [("If-None-Match", L.toStrict etag)]
                     }
-        req2 = req1 { checkStatus = \_ _ _ -> Nothing }
+        req2 = req1 { checkResponse = \_ _ -> return () }
```
Comment (sjakobi) [inline src/Network/HTTP/Download.hs]: @borsboom: In this case the we're handling the status codes manually afterwards, so I think it's still good to override the `checkResponse` method that's passed in.

**Comment by sjakobi** [comment]
@borsboom: I've added a second commit that should be more in line with your comment. I previously hadn't checked which requests were handled by `redownload`.

**Comment by borsboom** [comment]
LGTM!  Thanks!

**Pull request merged**